utils/src/bitmap/historical/bitmap.rs

```diff
@@ -0,0 +1,526 @@
+            .filter(|(chunk_index, _)| **chunk_index >= newer_pruned)
+        {
+            match change {
+                ChunkDiff::Modified(old_data) | ChunkDiff::Removed(old_data) => {
+                    // Both cases: chunk exists in target, just update its data
+                    newer_state.set_chunk_by_index(chunk_index, old_data);
+                }
+                ChunkDiff::Added => {
+                    // Chunk didn't exist in target - already handled by pop_to_length.
+                    // We can break here because there are no more modifications to apply.
+                    // Added can only occur after all Modified. If we encounter Added, we know
+                    // there are no Removed. (diff.chunk_diffs can't have both Added and Removed.)
+                    break;
+                }
+                ChunkDiff::Pruned(_) => {
+                    panic!("pruned chunk found at unexpected index {chunk_index}")
+                }
+            }
+        }
+
+        assert_eq!(newer_state.pruned_chunks(), target_pruned);
+        assert_eq!(newer_state.len(), target_len);
+    }
+
+    /// Check if a commit exists.
+    pub fn commit_exists(&self, commit_number: u64) -> bool {
+        self.commits.contains_key(&commit_number)
+    }
+
+    /// Get an iterator over all commit numbers in ascending order.
+    pub fn commits(&self) -> impl Iterator<Item = u64> + '_ {
+        self.commits.keys().copied()
+    }
+
+    /// Get the latest commit number, if any commits exist.
+    pub fn latest_commit(&self) -> Option<u64> {
+        self.commits.keys().next_back().copied()
+    }
+
+    /// Get the earliest commit number, if any commits exist.
+    pub fn earliest_commit(&self) -> Option<u64> {
+        self.commits.keys().next().copied()
+    }
+
+    /// Get a reference to the current bitmap state.
+    pub fn current(&self) -> &Prunable<N> {
+        &self.current
+    }
+
+    /// Number of bits in the current bitmap.
+    #[inline]
+    pub fn len(&self) -> u64 {
+        self.current.len()
+    }
+
+    /// Returns true if the current bitmap is empty.
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.current.is_empty()
+    }
+
+    /// Get the value of a bit in the current bitmap.
+    #[inline]
+    pub fn get_bit(&self, bit: u64) -> bool {
+        self.current.get_bit(bit)
+    }
+
+    /// Get the chunk containing a bit in the current bitmap.
+    #[inline]
+    pub fn get_chunk_containing(&self, bit: u64) -> &[u8; N] {
+        self.current.get_chunk_containing(bit)
+    }
+
+    /// Number of pruned chunks in the current bitmap.
+    #[inline]
+    pub fn pruned_chunks(&self) -> usize {
+        self.current.pruned_chunks()
+    }
+
+    /// Remove all commits with numbers below the commit number.
+    ///
+    /// Returns the number of commits removed.
+    pub fn prune_commits_before(&mut self, commit_number: u64) -> usize {
+        let keys_to_remove: Vec<u64> = self
```

Comment:
Can you more concisely (and possibly more efficiently) use something like self.commit.split_off here?

Reply:
split_off looks pretty efficient to me actually: https://doc.rust-lang.org/src/alloc/collections/btree/map.rs.html#1375-1378